File: python/src/falcon_core/communications/messages/measurement_request.py
from __future__ import annotations
from typing import Any, List, Dict, Tuple, Optional
from falcon_core._capi.measurement_request import MeasurementRequest as _CMeasurementRequest
from falcon_core.math.domains.labelled_domain import LabelledDomain
from falcon_core.generic.list import List
from falcon_core.generic.map import Map
from falcon_core.instrument_interfaces.names.ports import Ports

class MeasurementRequest:
    """Python wrapper for MeasurementRequest."""

    def __init__(self, c_obj):
        self._c = c_obj

    @classmethod
    def _from_capi(cls, c_obj):
        if c_obj is None:
            return None
        return cls(c_obj)

    @classmethod
    def from_json(cls, json: str) -> MeasurementRequest:
        return cls(_CMeasurementRequest.from_json(json))

    @classmethod
    def new(cls, message: str, measurement_name: str, waveforms: List, getters: Ports, meter_transforms: Map, time_domain: LabelledDomain) -> MeasurementRequest:
        obj = cls(_CMeasurementRequest.new(message, measurement_name, waveforms._c if waveforms is not None else None, getters._c if getters is not None else None, meter_transforms._c if meter_transforms is not None else None, time_domain._c if time_domain is not None else None))
        obj._ref_waveforms = waveforms  # Keep reference alive
        obj._ref_getters = getters  # Keep reference alive
        obj._ref_meter_transforms = meter_transforms  # Keep reference alive
        obj._ref_time_domain = time_domain  # Keep reference alive
        return obj

    def copy(self, ) -> MeasurementRequest:
        ret = self._c.copy()
        return MeasurementRequest._from_capi(ret)

    def equal(self, other: MeasurementRequest) -> None:
        ret = self._c.equal(other._c if other is not None else None)
        return ret

    def not_equal(self, other: MeasurementRequest) -> None:
        ret = self._c.not_equal(other._c if other is not None else None)
        return ret

    def to_json(self, ) -> str:
        ret = self._c.to_json()
        return ret

    def measurement_name(self, ) -> str:
        ret = self._c.measurement_name()
        return ret

    def getters(self, ) -> Ports:
        ret = self._c.getters()
        if ret is None: return None
        return Ports._from_capi(ret)

    def waveforms(self, ) -> List:
        ret = self._c.waveforms()
        if ret is None: return None
        return List(ret)

    def meter_transforms(self, ) -> Map:
        ret = self._c.meter_transforms()
        if ret is None: return None
        return Map(ret)

    def time_domain(self, ) -> LabelledDomain:
        ret = self._c.time_domain()
        if ret is None: return None
        return LabelledDomain._from_capi(ret)

    def message(self, ) -> str:
        ret = self._c.message()
        return ret

    def __hash__(self):
        """Hash based on JSON representation"""
        return hash(self.to_json())

    def __eq__(self, other):
        """Operator overload for =="""
        if not isinstance(other, MeasurementRequest):
            return NotImplemented
        return self.equal(other)

    def __ne__(self, other):
        """Operator overload for !="""
        if not isinstance(other, MeasurementRequest):
            return NotImplemented
        return self.not_equal(other)

File: ocaml/_opam/.opam-switch/sources/falcon_core/src/falcon_core/communications/messages/measurementrequest.ml
open Ctypes
open Capi_bindings
open ErrorHandling

open Falcon_core.Generic
open Falcon_core.Instrument_interfaces.Names
open Falcon_core.Math.Domains

class c_measurementrequest (h : unit ptr) = object(self)
  val raw_val = h
  method raw = raw_val
  initializer Gc.finalise (fun _ ->
    Capi_bindings.measurementrequest_destroy raw_val;
    ErrorHandling.raise_if_error ()
  ) self
end

module MeasurementRequest = struct
  type t = c_measurementrequest

  let copy (handle : t) : t =
    ErrorHandling.read handle (fun () ->
      let ptr = Capi_bindings.measurementrequest_copy handle#raw in
      ErrorHandling.raise_if_error ();
      new c_measurementrequest ptr
    )

  let fromjson (json : string) : t =
    let ptr = Capi_bindings.measurementrequest_from_json_string (Capi_bindings.string_wrap json) in
    ErrorHandling.raise_if_error ();
    new c_measurementrequest ptr

  let make (message : string) (measurement_name : string) (waveforms : ListWaveform.t) (getters : Ports.t) (meter_transforms : MapInstrumentPortPortTransform.t) (time_domain : LabelledDomain.t) : t =
    ErrorHandling.multi_read [waveforms; getters; meter_transforms; time_domain] (fun () ->
      let ptr = Capi_bindings.measurementrequest_create (Capi_bindings.string_wrap message) (Capi_bindings.string_wrap measurement_name) waveforms#raw getters#raw meter_transforms#raw time_domain#raw in
      ErrorHandling.raise_if_error ();
      new c_measurementrequest ptr
    )

  let equal (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measurementrequest_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let notEqual (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.measurementrequest_not_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let toJsonString (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measurementrequest_to_json_string handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let measurementName (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measurementrequest_measurement_name handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let getters (handle : t) : Ports.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measurementrequest_getters handle#raw in
      ErrorHandling.raise_if_error ();
      new c_ports result
    )

  let waveforms (handle : t) : ListWaveform.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measurementrequest_waveforms handle#raw in
      ErrorHandling.raise_if_error ();
      new c_listwaveform result
    )

  let meterTransforms (handle : t) : MapInstrumentPortPortTransform.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measurementrequest_meter_transforms handle#raw in
      ErrorHandling.raise_if_error ();
      new c_mapinstrumentportporttransform result
    )

  let timeDomain (handle : t) : LabelledDomain.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measurementrequest_time_domain handle#raw in
      ErrorHandling.raise_if_error ();
      new c_labelleddomain result
    )

  let message (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.measurementrequest_message handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

end